docs/book/user-guide/advanced-guide/best-practices/unit_testing.md

```diff
@@ -0,0 +1,779 @@
+- **Execution Status**: `[100%]` signifies that all test cases have completed. The dots represent passed tests (`.`).
+- **Summary**: Indicates the test outcome—`2 passed` test cases—in a specified duration (`0.04s`).
+
+### Understanding Results
+
+- **Passed**: The number of test cases that ran successfully.
+- **Failed/Skipped**: Any test cases that did not pass or were skipped during execution.
+- **Execution Time**: The time taken to execute the entire test suite.
+
+## Some tips and best practices to follow while writing unit testing for ZenML
+
+### Isolation:
+Ensure each test is independent and doesn't rely on external dependencies or other test cases for its execution. This prevents cascading failures and makes debugging easier.
+
+```Python
+import unittest
+from zenml.core.pipelines.training_pipeline import TrainingPipeline
+from zenml.core.steps.evaluator.tf_evaluator import TFEvaluator
+from zenml.core.steps.split.random_split import RandomSplit
+from zenml.core.steps.preprocesser.standard_scaler import StandardScaler
+from zenml.core.steps.trainer.tf_trainer import TFTrainer
+from unittest.mock import patch, MagicMock
+
+class TestZenMLIsolation(unittest.TestCase):
+    
+    def setUp(self):
+        # Initialize your ZenML objects
+        self.datasource_mock = MagicMock()
+        self.split_mock = MagicMock()
+        self.preprocess_mock = MagicMock()
+        self.trainer_mock = MagicMock()
+        self.evaluator_mock = MagicMock()
+        
+    def test_pipeline_steps_isolation(self):
+        # Create a training pipeline
+        training_pipeline = TrainingPipeline(name='my_training_pipeline',
+                                             datasource=self.datasource_mock,
+                                             steps=[self.split_mock,
+                                                    self.preprocess_mock,
+                                                    self.trainer_mock,
+                                                    self.evaluator_mock])
+        
+        # Set isolation for each step in the pipeline
+        for step in training_pipeline.steps:
+            self.assertTrue(step.isolation, msg=f"{step} is not isolated.")
+            
+    def test_step_isolation(self):
+        # Create individual steps
+        split = RandomSplit()
+        preprocess = StandardScaler()
+        trainer = TFTrainer()
+        evaluator = TFEvaluator()
+        
+        # Set isolation for each step individually
+        split.isolation = True
+        preprocess.isolation = True
+        trainer.isolation = True
+        evaluator.isolation = True
+        
+        self.assertTrue(split.isolation, msg="RandomSplit step is not isolated.")
+        self.assertTrue(preprocess.isolation, msg="StandardScaler step is not isolated.")
+        self.assertTrue(trainer.isolation, msg="TFTrainer step is not isolated.")
+        self.assertTrue(evaluator.isolation, msg="TFEvaluator step is not isolated.")
```

Comment:
This is wrong syntax for ZenML pipelines, datasource and steps are not an input, and overall this seems like old GPT generated ;-)

Reply:
Oops, Got caught  .sorry for that I need to grow / change my habit . I'll fix that out :)